heft-plugins/heft-webpack5-plugin/src/Webpack5Plugin.ts

```diff
@@ -148,131 +164,241 @@ export default class Webpack5Plugin implements IHeftTaskPlugin<IWebpack5PluginOp
+            true,
+            taskSession.logger.terminal
           );
-        } catch (e) {
-          logger.emitError(e as Error);
+          devServerOptions = {
+            ...devServerOptions,
+            server: {
+              type: 'https',
+              options: {
+                key: certificate.pemKey,
+                cert: certificate.pemCertificate
+              }
+            }
+          };
         }
+
+        // Since the webpack-dev-server does not return infrastructure errors via a callback like
+        // compiler.watch(...), we will need to intercept them and log them ourselves.
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        compiler.hooks.infrastructureLog.tap(PLUGIN_NAME, (name: string, type: string, args: any[]) => {
+          if (name === WEBPACK_DEV_MIDDLEWARE_PACKAGE_NAME && type === 'error') {
+            const error: Error | undefined = args[0];
+            if (error) {
+              taskSession.logger.emitError(error);
+            }
+          }
+          return true;
+        });
+
+        // The webpack-dev-server package has a design flaw, where merely loading its package will set the
+        // WEBPACK_DEV_SERVER environment variable -- even if no APIs are accessed. This environment variable
+        // causes incorrect behavior if Heft is not running in serve mode. Thus, we need to be careful to call
+        // require() only if Heft is in serve mode.
+        taskSession.logger.terminal.writeLine('Starting webpack-dev-server');
+        const WebpackDevServer: typeof TWebpackDevServer = (await import(WEBPACK_DEV_SERVER_PACKAGE_NAME))
+          .default;
+        const webpackDevServer: TWebpackDevServer = new WebpackDevServer(devServerOptions, compiler);
+        await webpackDevServer.start();
       } else {
-        try {
-          stats = await LegacyAdapters.convertCallbackToPromise(
-            (compiler as TWebpack.Compiler).run.bind(compiler)
-          );
-          await LegacyAdapters.convertCallbackToPromise(compiler.close.bind(compiler));
-        } catch (e) {
-          logger.emitError(e as Error);
-        }
+        // Create the watcher. Compilation will start immediately after invoking watch().
+        taskSession.logger.terminal.writeLine('Starting Webpack watcher');
+        compiler.watch({}, (error?: Error | null) => {
+          if (error) {
+            taskSession.logger.emitError(error);
+          }
+        });
       }
 
-      if (stats) {
-        this._emitErrors(logger, stats, heftConfiguration.buildFolderPath);
-        if (this.accessor.hooks.onEmitStats.isUsed()) {
-          await this.accessor.hooks.onEmitStats.promise(stats);
-        }
-      }
+      // Store the watchers to be used for suspend/resume
+      this._webpackWatchers = (compiler as ExtendedMultiCompiler).compilers?.map(
+        (compiler: ExtendedCompiler) => compiler.watching
+      ) ?? [(compiler as ExtendedCompiler).watching];
```

Comment:
Same syntax suggestion of `((compiler as ExtendedMultiCompiler).compilers ?? [(compiler as ExtendedCompiler)]).map(compiler => compiler.watching)`